### PR TITLE
Profiles, technical part

### PR DESCRIPTION
**Pull request by badboy**
Based on #165 (should merge cleanly afterwards, but I can also rebase it if necessary before this is merged)

[Rendered](https://badboy.github.io/firefox-data-docs-pcd/concepts/profile/profile_creation.html) (includes #165)
[My changes only](https://github.com/badboy/firefox-data-docs/compare/df51a0fd4772de2841f01de3d1506d9e9b0f35d2...badboy:pcd-creation-v2)

### CODE REVIEW

File: high_level/profile/profile_creation.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+A profile can also be reset or deleted by the user.
+The user can opt-out of or opt-in to sending Telemetry data.
+
+These different cases are described in more detail in the following sections.
+
+### Profile Creation
+
+#### First use
+
+When Firefox is opened for the first time after a fresh install, without any prior Firefox profile on disk, it will create a new profile.
```
Comment (SuYoungHong) [inline high_level/profile/profile_creation.md]: "When Firefox is opened for the first time after a fresh install, without any prior Firefox profile on disk"

"When Firefox is opened for the first time after a fresh install, without any prior Firefox profile on disk **visible to Firefox**"

- Firefox has some specific directories it checks for Profiles, right? Where it looks for / keeps profiles by default. Might be good to call that out.

File: high_level/profile/profile_creation.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+Firefox chooses a random name, create the profile directory and marks the new profile as default for subsequent starts of Firefox.
+
+#### Default or Profile Manager creation
+
+On subsequent starts, Firefox tries to find the profile directory and load the profile data (bookmarks, addons, etc.).
+If not otherwise specified it uses the default (or last used) profile, finds the profile directory and starts loading files from there.
+
```
Comment (SuYoungHong) [inline high_level/profile/profile_creation.md]: The above part seems out of place with this section. It's talking about how FF looks for profiles and decides which profile to use, while this bigger section is about how profiles are created. 

might be a good idea to break out a separate section for what a "Default" profile is and it's role

File: high_level/profile/profile_creation.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+The user can create a new profile through the profile manager.
+This can either be done on `about:profiles` or by starting Firefox with the `--ProfileManager` flag.
+The profile manager will ask for a name for the profile and picks a new directory for it.
+The directory can be changed by the user to another new or even an existing directory.
+If that directory does not exist, it is created.
+Additional files for bookmarks are placed inside this directory.
+
```
Comment (SuYoungHong) [inline high_level/profile/profile_creation.md]: The wording is a bit confusing in this section. Would be good to simplify, like: 

"The profile manager allows the user to create a new profile from an existing directory (in which case any files will be included) or from scratch (in which case the directory will be created."

File: high_level/profile/profile_creation.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+#### Command-line start
+
+Firefox can be started on the command line with a path to a profile directory: `firefox --profile path/to/directory`.
+In that case the passed profile directory is, if not already existing, created. No files inside that directory are written at launch.
+
```
Comment (SuYoungHong) [inline high_level/profile/profile_creation.md]: sorry, confused by the statement "No files inside that directory are written at launch."

File: high_level/profile/profile_creation.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+### Profile Reset
+
+TODO
+
+### Profile Deletion
+
+A profile can be deleted by simply removing the profile directory.
+We will never know about a deletion. We simply won't see that profile turning up in Telemetry data anymore.
+
```
Comment (SuYoungHong) [inline high_level/profile/profile_creation.md]: Might want to include how, just so it's explicitly clear. Profile deletion via: 
* Profile manager
* Deleting the profile directory
* Deleting Firefox does NOT delete profile directory

File: high_level/profile/profile_creation.md (hunk 4 above)
Comment (SuYoungHong) [reply high_level/profile/profile_creation.md]: Also, one important distinction that we might want to mention is how with this method, this profile will **not persist** in the sense that the directory will remain but it won't be considered a 'profile' by Firefox (unless you run FF with this command to the same directory every time).

The directory you use as a 'profile' in this case will not be picked up by Profile Manager, whereas a profile created by default on first run of FF, or via Profile manager will be visible and available to profile manager afterwards

File: high_level/profile/profile_creation.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+## Profile Creation Date
+
+The *profile creation date* is the assumed date of initial profile creation.
+However it proofed to be not reliable for all cases.
+There are multiple ways this date is determined.
+
+### During Profile Creation
+
+When a profile is created explicitly (default profile on first launch, new profile from Profile Manager or the `--createprofile` on the command line), the profile directory is created and a `times.json` containing a timestamp of the current time is stored inside that profile directory[^1].
+It is read at later times when the profile creation date is used.
```
Comment (SuYoungHong) [inline high_level/profile/profile_creation.md]: this category of profile creation (created explicitly) seems super useful, would be great if we could define it in the profile creation section above:

created explicity
* default profile on first launch
* new profile from Profile Manager
* --createprofile command

not created explicitly (?)
* --profile

Comment (SuYoungHong) [reply high_level/profile/profile_creation.md]: it would help understanding the flowchart a lot

File: high_level/profile/profile_creation.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+In order to understand the behavior of users and base analysis on things like the profile creation date,
+it is essential to understand how a profile is created and identified by the browser.
+
+A profile can also be reset or deleted by the user.
+The user can opt-out of or opt-in to sending Telemetry data.
+
+These different cases are described in more detail in the following sections.
+
+### Profile Creation
```
Comment (georgf) [inline high_level/profile/profile_creation.md]: Firefox Developer edition [automatically sets up a separate profile](https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Developer_Edition) for you, so you can easily run it in parallel with Firefox Release.
Can we call this out as well?

**Comment by georgf** [comment]
Chris, can you take a look at the technical correctness?
This is not 100% complete yet (aiming to publish rather than wait for 100%), some sections will be filled in later.

**Comment by badboy** [comment]
I addressed @SuYoungHong's and @georgf's initial feedback. Turning it into explicit/implicit profile creation makes it much easier to follow (and refer back) to the ways it's done.

File: concepts/profile/profile_creation.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+When the user opts out, Telemetry sends one ["optout" ping](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/data/optout-ping.html), containing an empty payload.
+The local `clientID` is reset to a fixed value.
+
+When a user opts into sending Telemetry data, a new `clientID` is generated and used in subsequent pings.
+The profile itself and the profile creation date are unaffected by this.
+
+## Profile Creation Date
+
+The *profile creation date* is the assumed date of initial profile creation.
+However it proofed to be not reliable for all cases.
```
[the remark this answers is not in the document]
Comment (badboy) [reply concepts/profile/profile_creation.md]: Thanks. English is hard.

File: concepts/profile/profilehistory.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+![](images/profile-history/example-starting.png)
+
+As you can see, this profile starts with a ping where `profile_subsession_counter = 1` and `previous_subsession_id is none`.
+
+#### History Has No Beginning
+
+![](images/profile-history/example-midhistory.png)
+
+In this example, the profile simply appears in our data mid-history, with presumably the 25th subsession in it's history. It's previous history is a mystery.
```
[the remark this answers is not in the document]
Comment (badboy) [reply concepts/profile/profilehistory.md]: I mark this as a follow-up, so we can properly phrase that.

[the remark this answers is not in the document]
Comment (badboy) [reply concepts/profile/profilehistory.md]: #171

File: concepts/profile/profilehistory.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+* Profile begins
+
+![](images/profile-history/example-restart-1.png)
+
+* Profile restarts
+
+![](images/profile-history/example-restart-2.png) 
+
+Here, we see the profile start their history normally, but then they begin a new, totally unconnected branch with a starting ping that is **not** the same as the original starting ping (different `subsession_id`s).
```
[the remark this answers is not in the document]
Comment (badboy) [reply concepts/profile/profilehistory.md]: The included image shows that the profile_reset_date is None.
I'll file a follow-up to keep track of that.

[the remark this answers is not in the document]
Comment (badboy) [reply concepts/profile/profilehistory.md]: https://github.com/mozilla/firefox-data-docs/issues/172

File: concepts/profile/realworldusage.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+* 5: Firefox binary / installer
+	- The downloaded binary package or stub installer which installs Firefox into the hard drive. Users can get these from our website or one of our managed properties, but they can also acquire these from 3rd party sources as well. 
+	- Our website is instrumented with Google Analytics to track download numbers, but other properties (FTP) and 3rd party sources are not. Google Analytics data is not directly connected to Telemetry data. 
+	- A user can produce multiple installations from a single Firefox binary / installer. For example, if a user copies it to a USB stick or keeps it in cloud storage, they could install Firefox on multiple machines from a single binary / installer. 
+* 6: Firefox installation
+	- The installed Firefox program on a given disk image. 
+	- Since Firefox is usually installed in a system directory, the single installation of Firefox will be shared by all the OS user profiles in the disk image. 
+	- Stub installers are instrumented with pings to report new install counts, however, full binaries are not.
```
[the remark this answers is not in the document]
Comment (badboy) [reply concepts/profile/realworldusage.md]: https://github.com/mozilla/firefox-data-docs/issues/173

File: concepts/profile/realworldusage.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+In this case, depending on the state of the disk image when it was copied, we could see multiple client ID's for each user+machine, or we could see all the user+machines sharing the same client ID. 
+
+If the disk image was copied after a Firefox profile was created, then the old user+machine and new user+machine will share the same client ID, and be submitting pings to us concurrently. 
+
+If the disk image was copied after the Firefox installation but before an initial Firefox profile was created, then each user+machine will get their own Firefox profile and client ID when they run Firefox for the first time. 
+
+As with the Multi-Profile User and Shared Computer case, even though there could be multiple Firefox profiles in this use case, there will only be one download and install event.  
+
+#### Migrations
+
```
[the remark this answers is not in the document]
Comment (badboy) [reply concepts/profile/realworldusage.md]: https://github.com/mozilla/firefox-data-docs/issues/174

**Comment by badboy** [comment]
Addressed all comments regarding profile creation here, the other one are included in #165.

This is now only blocked on getting #165 merged and then needs a rebase before merge.

**Comment by badboy** [comment]
@SuYoungHong and @chutten were okay with the new phrasing of managed/unmanaged. I'm now gonna land this.

**Pull request merged**